subsys/bluetooth/services/cts.c

```diff
@@ -0,0 +1,192 @@
+	 * 'Date Time' - characteristic contains fields for:
+	 * year, month, day, hours, minutes and seconds.
+	 */
+	ct_time->year = sys_cpu_to_le16(date_time.tm_year);
+	ct_time->mon = date_time.tm_mon + 1; /* months starting from 1 */
+	ct_time->mday = date_time.tm_mday;   /* Day of month */
+	ct_time->hours = date_time.tm_hour;  /* hours */
+	ct_time->min = date_time.tm_min;     /* minutes */
+	ct_time->sec = date_time.tm_sec;     /* seconds */
+	/* day of week starting from 1-monday, 7-sunday */
+	ct_time->wday = date_time.tm_wday;
+	if (ct_time->wday == 0) {
+		ct_time->wday = 7; /* sunday is represented as 7 */
+	}
+	return 0;
+}
+
+#endif
+
+static void ct_ccc_cfg_changed(const struct bt_gatt_attr *attr, uint16_t value)
+{
+	bool notif_enabled = (value == BT_GATT_CCC_NOTIFY);
+
+	LOG_INF("CTS Notifications %s", notif_enabled ? "enabled" : "disabled");
+
+	if (_cb->notification_changed) {
+		_cb->notification_changed(notif_enabled);
+	}
+}
+
+static ssize_t read_ct(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+		       uint16_t len, uint16_t offset)
+{
+	int err;
+	struct bt_cts_time_format ct_time;
+
+	if (_cb->fill_current_cts_time == NULL) {
+		return BT_GATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES);
+	}
+
+	err = _cb->fill_current_cts_time(&ct_time);
+	ct_time.reason = BT_CTS_UPDATE_REASON_UNKNOWN;
+
+	if (!err) {
+		return bt_gatt_attr_read(conn, attr, buf, len, offset, &ct_time, sizeof(ct_time));
+	} else {
+		return BT_GATT_ERR(BT_ATT_ERR_OUT_OF_RANGE);
+	}
+}
+
+static ssize_t write_ct(struct bt_conn *conn, const struct bt_gatt_attr *attr, const void *buf,
+			uint16_t len, uint16_t offset, uint8_t flags)
+{
+	int err;
+	struct bt_cts_time_format ct_time;
+
+	if (_cb->cts_time_write == NULL) {
+		return BT_GATT_ERR(BT_ATT_ERR_INSUFFICIENT_RESOURCES);
+	}
+
+	if ((offset != 0) || (offset + len != sizeof(ct_time))) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	memcpy(&ct_time, buf, sizeof(ct_time));
+	err = _cb->cts_time_write(&ct_time);
+	if (err) {
+		return BT_GATT_ERR(BT_CTS_ATT_ERR_VALUES_IGNORED);
+	}
+
+	err = bt_cts_send_notification(BT_CTS_UPDATE_REASON_MANUAL);
+	if (err) {
+		return BT_GATT_ERR(BT_ATT_ERR_UNLIKELY);
+	}
+
+	return len;
+}
+
+/* Current Time Service Declaration */
+BT_GATT_SERVICE_DEFINE(cts_svc, BT_GATT_PRIMARY_SERVICE(BT_UUID_CTS),
+		       BT_GATT_CHARACTERISTIC(BT_UUID_CTS_CURRENT_TIME,
+					      BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE |
+						      BT_GATT_CHRC_NOTIFY,
+					      BT_GATT_PERM_READ | BT_GATT_PERM_WRITE, read_ct,
+					      write_ct, NULL),
+		       BT_GATT_CCC(ct_ccc_cfg_changed, BT_GATT_PERM_READ | BT_GATT_PERM_WRITE));
+
+int bt_cts_init(const struct bt_cts_cb *cb)
+{
+	__ASSERT(cb == NULL, "Current Time service need valid `struct bt_cts_cb` callback");
+	_cb = cb;
+	return 0;
+}
+
+int bt_cts_send_notification(uint8_t reason)
+{
+	int err;
+	struct bt_cts_time_format ct_time;
+
+	if (_cb->fill_current_cts_time == NULL) {
+		return -EFAULT;
+	}
```

Comment:
If it's invalid to register a callback struct with this member set to NULL, why don't you check for it in `bt_cts_init()` and return `-EINVAL` from there? Here in `bt_cts_send_notification()` it seems like it should at most be an `__ASSERT` (if anything at all)